File: pset6/credit/credit.py
from cs50 import get_string
import re


def main():
    while True:
        card = get_string("Number: ")
        if re.match(r"\d+", card):
            break

    if validation(card):
        if re.match(r"^5[12345]", card) and re.match(r"\d{16}", card):
            print("MASTERCARD")

        elif re.match(r"^3[47]", card) and re.match(r"\d{15}", card):
            print("AMEX")

        elif re.match(r"^4", card) and re.match(r"\d{13,17}", card):
            print("VISA")

        else:
            print("INVALID")

    else:
        print("INVALID")


def validation(card):
    sum1 = 0
    sum2 = 0

    for c in range(len(card) - 2, -1, -2):
        if (int(card[c]) * 2) > 9:
            sum1 += int(str(int(card[c]) * 2)[0])
            sum1 += int(str(int(card[c]) * 2)[1])
        else:
            sum1 += int(card[c]) * 2

        sum2 += int(card[c + 1])

    if len(card) % 2 == 1:
        sum2 += int(card[0])

    total = sum1 + sum2
    return (total % 10 == 0)


main()

File: week1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool validation(long num);

int main(void)
{
    long card = get_long("Number: ");

    if (validation(card))
    {
        int lng;
        for (lng = 1; card > 100; lng++)
        {
            card = (card - (card % 10)) / 10;
        }

        if ((card == 51 || card == 52 || card == 53 || card == 54 || card == 55) && lng > 14)
        {
            printf("MASTERCARD\n");
        }
        else if ((card == 34 || card == 37) && lng > 13)
        {
            printf("AMEX\n");
        }
        else if ((card - (card % 10)) / 10 == 4 && lng > 11)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

bool validation(long num)
{
    int sum1 = 0;
    int sum2 = 0;

    int j;
    long i;

    for (i = num, j = 1; i > 0; i = (i - (i % 10)) / 10, j++)
    {
        if (j % 2 == 0)
        {
            int tmp = ((i % 10) * 2);
            if (tmp >= 10)
            {
                sum1 = sum1 + (tmp % 10) + (((tmp - (tmp % 10)) / 10) % 10);
            }
            else
            {
                sum1 = sum1 + tmp;
            }
        }
        else
        {
            sum2 = sum2 + (i % 10);
        }
    }

    return (sum1 + sum2) % 10 == 0;
}